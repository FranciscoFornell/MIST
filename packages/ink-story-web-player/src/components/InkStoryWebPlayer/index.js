import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import InkHeader from '../InkHeader';
import InkLines from '../InkLines';
import InkChoices from '../InkChoices';
import InkButton from '../InkButton';
import { pick } from 'mist-utils';
import InkStoryManager from 'ink-story-manager';

/**
 * Ink Story Web Player
 * @component
 * @example
 * const props = {
 *   storyContent: {"inkVersion":19,"root":[[{"#":"title: InkStory component example"},{"#":"author: John Doe"},{"#":"defaultContinueText: Continue"},{"#":"continueMaximally"},"^Line 1","\n","^Line 2","\n","^Line 3","\n","ev","str","^Choice 1","/str","/ev",{"*":"0.c-0","flg":20},"ev","str","^Choice 2","/str","/ev",{"*":"0.c-1","flg":20},{"c-0":["\n","^You choosed 1","\n",{"->":"0.g-0"},{"#f":5}],"c-1":["\n","^You choosed 2","\n",{"->":"0.g-0"},{"#f":5}],"g-0":["^Line 4","\n","end",["done",{"#n":"g-1"}],null]}],"done",null],"listDefs":{}},
 * };
 *
 * return (<div style={{minHeight: '150px'}}>
 *   <InkStoryWebPlayer {...props} />
 * </div>);
 */
const InkStoryWebPlayer = ({ storyContent }) => {
  const [storyManager, setStoryManager] = useState();
  const [lines, setLines] = useState([]);
  const [canContinue, setCanContinue] = useState(true);
  const [choices, setChoices] = useState([]);

  useEffect(() => {
    setStoryManager(new InkStoryManager(storyContent));
  }, []);

  const updateState = (getter) => {
    const { historyLines, canContinue, currentChoices } = getter();

    setLines(historyLines);
    setCanContinue(canContinue);
    setChoices(currentChoices);
  };

  const continueStory = (choiceIndex) => {
    updateState(() => storyManager.continueStory(choiceIndex));
  };

  const rewind = () => {
    updateState(() => storyManager.rewind());
  };

  return (
    <React.Fragment>
      {storyManager && (
        <React.Fragment>
          <InkHeader {...pick(storyManager.globalTags, 'title', 'author')} />
          <InkLines lines={lines} />
          {storyManager.canRewind && (
            <InkButton label={'<<'} onClick={() => rewind()} />
          )}
          {canContinue && (
            <InkButton
              onClick={() => continueStory()}
              label={storyManager.globalTags.defaultContinueText}
            />
          )}
          {choices.length > 0 && (
            <InkChoices choices={choices} clickHandler={continueStory} />
          )}
        </React.Fragment>
      )}
    </React.Fragment>
  );
};

InkStoryWebPlayer.propTypes = {
  /**
   * Inkle's ink story in json format
   */
  storyContent: PropTypes.object.isRequired,
};

export default InkStoryWebPlayer;
