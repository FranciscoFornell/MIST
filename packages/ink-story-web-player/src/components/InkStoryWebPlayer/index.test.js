import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import InkStoryWebPlayer from './index';
import {
  setStoryGlobalTags,
  setStoryNextStepData,
  StoryManagerContinueMethod,
  StoryManagerRewindMethod,
  resetStoryInstance,
} from 'ink-story-manager';

jest.mock('ink-story-manager');

const renderComponent = () => render(<InkStoryWebPlayer storyContent={{}} />);

afterEach(() => {
  jest.clearAllMocks();
  resetStoryInstance();
});

afterAll(() => {
  jest.restoreAllMocks();
});

describe('InkStory component', () => {
  it('should render correctly the initial state', () => {
    setStoryGlobalTags({
      defaultContinueText: 'Continue',
      author: 'Author Name',
      title: 'Story Title',
    });
    expect(renderComponent().container.outerHTML).toMatchSnapshot();
  });

  describe('Whe the user clicks on', () => {
    describe('continue', () => {
      beforeEach(() => {
        setStoryGlobalTags({
          defaultContinueText: 'Continue',
          author: 'Author Name',
          title: 'Story Title',
        });
        setStoryNextStepData({
          canContinue: false,
          historyLines: ['Line 1', 'Line 2', 'Line 3'],
          currentChoices: [
            { text: 'Choice 1', index: 0 },
            { text: 'Choice 2', index: 1 },
          ],
        });
      });

      it('should make StoryManager to continue', () => {
        const { getByText } = renderComponent();
        fireEvent.click(getByText('Continue'));

        expect(StoryManagerContinueMethod).toHaveBeenCalledWith(undefined);
      });

      it('should render correctly with lines, choices, and rewind button', () => {
        const { getByText, container } = renderComponent();
        fireEvent.click(getByText('Continue'));

        expect(container.outerHTML).toMatchSnapshot();
      });
    });

    describe('choice button', () => {
      beforeEach(() => {
        setStoryGlobalTags({
          defaultContinueText: 'Continue',
          author: 'Author Name',
          title: 'Story Title',
        });
        setStoryNextStepData({
          canContinue: false,
          historyLines: ['Line 1', 'Line 2', 'Line 3'],
          currentChoices: [
            { text: 'Choice 1', index: 0 },
            { text: 'Choice 2', index: 1 },
            { text: 'Choice 3', index: 2 },
            { text: 'Choice 4', index: 3 },
          ],
        });
      });

      it('should make StoryManager to select the propper choice and continue', () => {
        const { getByText } = renderComponent();
        fireEvent.click(getByText('Continue'));
        fireEvent.click(getByText('Choice 4'));

        expect(StoryManagerContinueMethod).toHaveBeenCalledWith(3);
      });

      it('should render accordingly', () => {
        const { getByText, container } = renderComponent();
        fireEvent.click(getByText('Continue'));

        expect(container.outerHTML).toMatchSnapshot();
      });
    });

    describe('rewind button', () => {
      beforeEach(() => {
        setStoryGlobalTags({
          defaultContinueText: 'Continue',
          author: 'Author Name',
          title: 'Story Title',
        });
        setStoryNextStepData({
          canContinue: false,
          historyLines: ['Line 1', 'Line 2', 'Line 3'],
          currentChoices: [{ text: 'Continue', index: 0 }],
        });
      });
      it('should make the story manager to rewind story state and history', () => {
        const { getByText } = renderComponent();
        fireEvent.click(getByText('Continue'));
        fireEvent.click(getByText('<<'));
        expect(StoryManagerRewindMethod).toHaveBeenCalled();
      });
      it('should render with the previous story state and history', () => {
        const { getByText, container } = renderComponent();
        fireEvent.click(getByText('Continue'));
        const previousRender = container.outerHTML;
        setStoryNextStepData({
          canContinue: false,
          historyLines: ['Line 1', 'Line 2', 'Line 3', 'Line 4', 'Line 5'],
          currentChoices: [{ text: 'Continue', index: 0 }],
        });
        fireEvent.click(getByText('Continue'));
        const newRender = container.outerHTML;
        fireEvent.click(getByText('<<'));
        expect(container.outerHTML).not.toEqual(newRender);
        expect(container.outerHTML).toEqual(previousRender);
      });
    });
  });
});
