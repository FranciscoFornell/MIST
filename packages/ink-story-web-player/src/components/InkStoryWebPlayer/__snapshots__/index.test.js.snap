// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InkStory component Whe the user clicks on choice button should render accordingly 1`] = `
<div>
  <h2>
    Story Title
  </h2>
  <h4>
    Author Name
  </h4>
  <p>
    Line 1
  </p>
  <p>
    Line 2
  </p>
  <p>
    Line 3
  </p>
  <button>
    &lt;&lt;
  </button>
  <button>
    Choice 1
  </button>
  <button>
    Choice 2
  </button>
  <button>
    Choice 3
  </button>
  <button>
    Choice 4
  </button>
</div>
`;

exports[`InkStory component Whe the user clicks on continue should render correctly with lines, choices, and rewind button 1`] = `
<div>
  <h2>
    Story Title
  </h2>
  <h4>
    Author Name
  </h4>
  <p>
    Line 1
  </p>
  <p>
    Line 2
  </p>
  <p>
    Line 3
  </p>
  <button>
    &lt;&lt;
  </button>
  <button>
    Choice 1
  </button>
  <button>
    Choice 2
  </button>
</div>
`;

exports[`InkStory component should render correctly the initial state 1`] = `
<div>
  <h2>
    Story Title
  </h2>
  <h4>
    Author Name
  </h4>
  <button>
    Continue
  </button>
</div>
`;
